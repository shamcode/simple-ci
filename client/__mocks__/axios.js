const METHODS = [
    'get',
    'post',
    'put',
    'delete'
];

const FIXTURES = {};
const MOCKS = {};

function mockMethodFactory( method ) {
    return jest.fn().mockImplementation(
        ( url ) => {
            if ( !FIXTURES[ method ].hasOwnProperty( url ) ) {
                throw new Error( `Missing fixture for ${method.toUpperCase()} url = "${url}"` );
            }
            const { success, data } = FIXTURES[ method ][ url ];
            return Promise[ success ? 'resolve' : 'reject' ]( {
                data
            } );
        }
    );
}

METHODS.forEach(
    method => {
        MOCKS[ method ] = mockMethodFactory( method );
        FIXTURES[ method ] = {};
    }
);

const defaultMocksData = ( () => {
    const data = {};
    data.chain = {
        id: 2,
        name: 'Test chain',
        command: 'echo 1'
    };
    data.project = {
        id: 1,
        name: 'Test',
        cwd: '/tmp/',
        chains: [ data.chain ]
    };
    return data;
} )();

export default {
    defaultMocksData: defaultMocksData,
    mocks: MOCKS,
    create: jest.fn().mockImplementation( () => {
        return {
            ...MOCKS,
            interceptors: {
                request: {
                    use: () => {}
                },
                response: {
                    use: () => {}
                }
            }
        };
    } ),
    use( method, url, data, success = true ) {
        if ( !METHODS.includes( method ) ) {
            throw new Error( `Unknown method: "${method}"`);
        }
        FIXTURES[ method ][ url ] = {
            data,
            success
        };
        return this;
    },
    useDefaultMocks() {
        this
            .use( 'get', '/projects', [ defaultMocksData.project ] )
            .use( 'get', '/projects/1', defaultMocksData.project )
            .use( 'get',  '/chains/2', defaultMocksData.chain );
        return this;
    }
};