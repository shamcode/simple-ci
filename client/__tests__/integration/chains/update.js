import setup, { app } from '../helpers';
import axios from 'axios';
jest.mock( 'axios' );

beforeEach( () => {
    jest.resetModules();
    jest.clearAllMocks();
    setup();
} );

it( 'update project chain', async() => {
    expect.assertions( 9 );

    axios
        .useDefaultMocks()
        .use( 'put', '/chains/2', null );

    await app.start();
    await app.project.open();
    await app.chain.open();
    await app.click( '[href="projects/1/chains/2"]' );

    await app.waitRendering();
    app.checkBody();

    expect( document.querySelector( '[name="name"]' ).value ).toBe(
        axios.defaultMocksData.chain.name
    );
    expect( document.querySelector( '[name="command"]' ).value ).toBe(
        axios.defaultMocksData.chain.command
    );

    const formData = {
        name: 'Test chain name',
        command: 'echo 2'
    };
    app.form.fill( 'name', formData.name );
    app.form.fill( 'command', formData.command );
    await app.form.submit();

    expect( axios.mocks.put ).toHaveBeenCalledTimes( 1 );
    expect( axios.mocks.put.mock.calls[ 0 ][ 0 ] ).toBe( '/chains/2' );
    expect( Object.keys( axios.mocks.put.mock.calls[ 0 ][ 1 ] ) ).toEqual( [ 'name', 'command' ] );
    expect( axios.mocks.put.mock.calls[ 0 ][ 1 ].name ).toEqual( formData.name );
    expect( axios.mocks.put.mock.calls[ 0 ][ 1 ].command ).toEqual( formData.command );
    app.checkBody();
} );


