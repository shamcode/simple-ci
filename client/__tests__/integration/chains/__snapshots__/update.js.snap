// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update project chain 1`] = `
"<header>
  <a class=\\"logo\\" href=\\"\\"> Simple CI </a>
</header>
<!--Header-->
<main>
  <ul class=\\"breadcrumb\\">
    <li>
      <a href=\\"\\"> Projects </a>
    </li>
    <li>
      <a href=\\"projects/1\\">Test</a>
    </li>
    <!--if-->
    <!--if-->
    <li>Test chain</li>
  </ul>
  <!--BreadCrumb-->
  <!--breadcrumbs-->
  <!--if-->
  <!--ErrorsPanel-->
  <div class=\\"chain-info\\">
    <h3>Test chain</h3>
  </div>
  <div class=\\"project-chain-form\\">
    <form>
      <div class=\\"group\\">
        <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
        <label>Name</label>
      </div>
      <div class=\\"group\\">
        <textarea wrap=\\"off\\" style=\\"display: none\\" name=\\"command\\">echo 1</textarea>
        <div class=\\"code-flask\\">
          <div class=\\"codeflask codeflask--has-line-numbers\\">
            <textarea class=\\"codeflask__textarea codeflask__flatten\\" spellcheck=\\"false\\" autocapitalize=\\"off\\" autocomplete=\\"off\\" autocorrect=\\"off\\"></textarea><pre class=\\"codeflask__pre codeflask__flatten  language-html\\"><code class=\\"codeflask__code  language-html\\">echo 1</code></pre>
            <div class=\\"codeflask__lines\\"><span class=\\"codeflask__lines__line\\">1</span></div>
          </div>
        </div>
        <!--CodeFlask--><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
        <label>Command</label>
      </div>
      <div class=\\"group\\">
        <button type=\\"submit\\"><span>Save</span></button>
      </div>
    </form>
  </div>
  <!--ChainForm-->
  <!--Info-->
  <!--if-->
  <!--content-->
  <!--if-->
  <!--PageContent-->
</main>
<footer> Sham 2018
  (v 0.0.5) </footer>
<!--Footer-->"
`;

exports[`update project chain 2`] = `
"<header>
  <a class=\\"logo\\" href=\\"\\"> Simple CI </a>
</header>
<!--Header-->
<main>
  <ul class=\\"breadcrumb\\">
    <li>
      <a href=\\"\\"> Projects </a>
    </li>
    <li>
      <a href=\\"projects/1\\">Test</a>
    </li>
    <!--if-->
    <!--if-->
    <li>Test chain</li>
  </ul>
  <!--BreadCrumb-->
  <!--breadcrumbs-->
  <!--if-->
  <!--ErrorsPanel-->
  <div class=\\"chain-info\\">
    <h3>Test chain</h3>
  </div>
  <div class=\\"project-chain-form\\">
    <form>
      <div class=\\"group\\">
        <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
        <label>Name</label>
      </div>
      <div class=\\"group\\">
        <textarea wrap=\\"off\\" style=\\"display: none\\" name=\\"command\\">echo 1</textarea>
        <div class=\\"code-flask\\">
          <div class=\\"codeflask codeflask--has-line-numbers\\">
            <textarea class=\\"codeflask__textarea codeflask__flatten\\" spellcheck=\\"false\\" autocapitalize=\\"off\\" autocomplete=\\"off\\" autocorrect=\\"off\\"></textarea><pre class=\\"codeflask__pre codeflask__flatten  language-html\\"><code class=\\"codeflask__code  language-html\\">echo 1</code></pre>
            <div class=\\"codeflask__lines\\"><span class=\\"codeflask__lines__line\\">1</span></div>
          </div>
        </div>
        <!--CodeFlask--><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
        <label>Command</label>
      </div>
      <div class=\\"group\\">
        <button type=\\"submit\\"><span>Save</span></button>
      </div>
    </form>
  </div>
  <!--ChainForm-->
  <!--Info-->
  <!--if-->
  <!--content-->
  <!--if-->
  <!--PageContent-->
</main>
<footer> Sham 2018
  (v 0.0.5) </footer>
<!--Footer-->"
`;
